File: 0823-Binary Trees With Factors/0823-Binary Trees With Factors.cpp
class Solution {
public:
    int numFactoredBinaryTrees(vector<int>& A) {
        int MOD = 1000000007, n = A.size();
        sort(A.begin(), A.end());
        vector<long long> dp(n, 1);
        unordered_map<int, int> table;
        for (int i = 0; i < n; ++i) {
            table[A[i]] = i;
        }
        long long total = 0;
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < i; ++j) {
                if (A[i] % A[j] == 0) {
                    int num = A[i] / A[j];
                    if (table.count(num)) {        
                        dp[i] = (dp[i] + dp[j] * dp[table[num]]) % MOD;
                    }                    
                }
            }
            total = (total + dp[i]) % MOD;
        }
        return total;
    }
};

File: 0823-Binary Trees With Factors/0823-Binary Trees With Factors.py
class Solution:
    def numFactoredBinaryTrees(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        MOD = 1000000007
        n = len(A)
        A.sort()
        table = {x: i for i, x in enumerate(A)}
        dp = [1] * n
        total = 0
        for i in range(n):
            for j in range(i):
                if A[i] % A[j] == 0:
                    num = A[i] / A[j]
                    if num in table:
                        dp[i] += dp[j] * dp[table[num]]
            total += dp[i]
        return total % MOD
